aether-odk-module/aether/odk/api/xform_utils.py

```
@@ -16,265 +16,772 @@
+        workbook_dict=xform_dict,
+        form_name=name,
+        default_language=language,
+        warnings=[],
+    )
+    survey = builder.create_survey_element_from_dict(json_survey)
+    return survey.xml().toprettyxml(indent='  ')
 
-        {
-            'ZZZ': {
-                '@id': 'form-id',
-                '@version': 'v1,
-                ...
-                'choice_a': 'id_1',
-                'number_b': '1',
-                ...
-            }
-        }
 
-    Into:
+def __parse_xmlform(fp):
+    '''
+    Parses XML file content into an XML string. Checking that the content is a
+    valid XML.
+    '''
 
-        {
-            '@id': 'form-id',
-            '@version': 'v1,
-            ...
-            'choice_a': 'value_1',
-            'number_b': 1,
-            ...
+    content = fp.read().decode('utf-8')
+    # check that the file content is a valid XML
+    __parse_xml_to_dict(content)
+    # but return the untouched content if it does not raise any exception
+    return content
+
+
+def __parse_xml_to_dict(xml_content):
+    '''
+    Parses XML file content into an dictionary.
+    '''
+
+    root = ElementTree.fromstring(xml_content)
+    xml_dict = __etree_to_dict(root)
+    return xml_dict
+
+
+def __get_xform_instance(xform_dict, with_root=False):
+    '''
+    Extracts instance from xForm definition
+    '''
+
+    try:
+        instances = __wrap_as_list(xform_dict['h:html']['h:head']['model']['instance'])
+        for i in instances:
+            # the default instance is the only one without "id" attribute
+            if '@id' not in i:
+                instance = i
+                break
+
+        if isinstance(instance, dict):
+            if with_root:
+                return instance
+
+            # assumption: there is only one child (key)
+            key = list(instance.keys())[0]
+            return instance[key]
+
+    except Exception:
+        pass
+
+    raise TypeError('missing instance definition')
```

Maybe this would be a good place for an application-specific error? TypeError is very general, I'd expect to see that raised when python-related issues happen, such as `int({})`. And this is more related to the shape/content of the xml, if I'm not mistaken. 

Also, could we raise the new error inside the `Exception` block?
```
except KeyError:
    raise MissingInstanceError('...')
```